### PR TITLE
Implementation of Girsanov reweighting in deeptime MSM estimator

### PR DESCRIPTION
**Pull request by anyschaefer**
We extended the MSM estimator class, such that the transition counts are reweighted according to the reweighting factor trajectory. 

Joana-Lysiane Schäfer, Bettina G. Keller, Implementation of Girsanov reweighting in OpenMM and Deeptime

We demonstrated the correct functioning and error-free applicability of the newly implemented
functions using both a low-dimensional test system and a molecular system. 
The corresponding reference "Implementation of Girsanov reweighting in OpenMM and Deeptime" is currently under review and will be publicly available shortly. 
The required input files and setup for the evaluation of a reweighted Markov state model are available at [reweightingtools](https://github.com/bkellerlab/reweightingtools)

### CODE REVIEW

**Comment by clonker** [comment]
Super cool, thank you for your contribution! Do you think it is possible to add a sanity unit test to the new estimator? The failing tests are due to a tab/space mismatch. Let me know if you need support, I am happy to integrate this.

File: deeptime/markov/tools/estimation/api.py [hunk 1]
```diff
@@ -213,7 +214,83 @@ def effective_count_matrix(dtrajs, lag, average='row', mact=1.0, n_jobs=None, ca
     return sparse.effective_counts.effective_count_matrix(dtrajs, lag, average=average, mact=mact, n_jobs=n_jobs,
                                                           callback=callback)
 
+def girsanov_reweighted_count_matrix(dtraj, lag, reweighting_factors,
+                                     sliding=True, sparse_return=True, nstates=None):
+    r"""Generate a Girsanov reweighted count matrix from given microstate trajectory. :footcite:`donati2017girsanov`
```
Comment (clonker) [inline deeptime/markov/tools/estimation/api.py:219]: This will need an entry in [references.bib](docs/source/references.bib) to render correctly

Comment (anyschaefer) [reply deeptime/markov/tools/estimation/api.py:219]: we will add the reference as soon as it is published

**Comment by anyschaefer** [comment]
> Super cool, thank you for your contribution! Do you think it is possible to add a sanity unit test to the new estimator? The failing tests are due to a tab/space mismatch. Let me know if you need support, I am happy to integrate this.

Thank you, that's great! We added a sanity unit test for the new estimator and fixed the tap/space mismatch. When trying to run the tests with `pytest tests/` or reinstall the modified package we get import errors due to different `..._bindings.cpython-311-x86_64-linux-gnu.so` files. The problem concerns modules of deeptime that we have not modified, how to deal with this problem?

**Comment by clonker** [comment]
The bindings are part of the c++ module of deeptime, If all dependencies are available, you can set up a development environment with it using `python setup.py develop`. Otherwise you can just use `pip install .` and then use that module for testing. There are some tests failing right now in CI due to another part of deeptime, I have to fix that :) but generally, if your test runs through then that's fine for me.

**Comment by clonker** [comment]
If you update your branch against the current main, the errors unrelated to this PR should be gone. :)

File: deeptime/markov/_girsanov_reweighted_transition_counting.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        >>> print(reweighted_counts.count_matrix)
+        """ 
+        if count_mode == 'sliding':
+            reweighting_factors=(ensure_factors_list(reweighting_factors[0]),ensure_factors_list(reweighting_factors[1]))
+            count_matrix = msmest.girsanov_reweighted_count_matrix(dtrajs, lagtime, reweighting_factors, 
+                                                                   sliding=True, sparse_return=sparse)
+        elif count_mode in {'sample','effective','sliding-effective'}:
+            raise ValueError('Count mode {} is not compatible with the Girsanov reweighting estimator.'.format(count_mode))
+        else:
+            raise ValueError('Count mode {} is unknown.'.format(count_mode))
```
Comment (clonker) [inline deeptime/markov/_girsanov_reweighted_transition_counting.py:114]:
```suggestion
        else
            raise ValueError('Count mode {} is not compatible with the Girsanov reweighting estimator, only "sliding" supported.'.format(count_mode))
```

File: deeptime/markov/tools/estimation/sparse/count_matrix.py [hunk 3]
```diff
@@ -60,7 +67,26 @@ def count_matrix_coo2_mult(dtrajs, lag, sliding=True, sparse=True, nstates=None)
...
+        for g,M in zip(g_factors,M_factors):
+            if g.size > lag:
+                if sliding:
+                    m = M.cumsum()
+                    m[lag:] = m[lag:] - m[:len(m)-lag]
+                    m = m[(lag):]
+                    m = np.exp(-m)
+                    factors.append(g[0:-lag]*m)
+                else:
+                    raise NotImplementedError('Only the sliding scheme is implemented.')
```
Comment (clonker) [inline deeptime/markov/tools/estimation/sparse/count_matrix.py:85]: I would expect that this is checked sooner, eg, right below the `elif` in line 73.

File: deeptime/markov/tools/estimation/sparse/count_matrix.py [hunk 4]
```diff
@@ -60,7 +67,26 @@ def count_matrix_coo2_mult(dtrajs, lag, sliding=True, sparse=True, nstates=None)
...
+        factors = np.concatenate(factors)
+        data = factors
+    else:
+        raise NotImplementedError('An input format other than a tuple (g,M) for the reweighting factors is not implemented.')
```
Comment (clonker) [inline deeptime/markov/tools/estimation/sparse/count_matrix.py:89]: could be covered by tests

File: deeptime/util/types.py [hunk 5]
```diff
@@ -126,6 +126,11 @@ def ensure_dtraj_list(dtrajs):
         return [ensure_integer_array(dtrajs)]
     return [ensure_integer_array(t) for t in dtrajs]
 
+def ensure_factors_list(factors):
+    """Makes sure that g_factors and M_factors are a list of reweighting factors (array of int)"""
+    if len(factors) > 0 and isinstance(factors[0], numbers.Real):
+        return [ensure_integer_array(factors)]
+    return [ensure_integer_array(t) for t in factors]
```
Comment (clonker) [inline deeptime/util/types.py:133]: should be covered by tests as list of factor lists

File: tests/markov/test_girsanov_reweighted_transition_counting.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        estimator = GirsanovReweightingEstimator(lagtime=2, count_mode="sliding-effective")
+        with self.assertRaises(ValueError):
+            estimator.fit(dtraj[:-1], reweighting_factors=_reweighting).fetch_model()
+
+    def test_effective_counting(self):
+        dtraj = np.array([0, 0, 0, 0, 1, 1, 0, 1])
+        _reweighting = (np.array([1, 1, 1, 1, 1, 1, 1, 1]),np.array([1, 1, 1, 1, 1, 1, 1, 1]))
+        estimator = GirsanovReweightingEstimator(lagtime=2, count_mode="effective")
+        with self.assertRaises(ValueError):
+            estimator.fit(dtraj[:-1], reweighting_factors=_reweighting).fetch_model()
```
Comment (clonker) [inline tests/markov/test_girsanov_reweighted_transition_counting.py:58]: please add blank line at the end of file :)

File: tests/markov/test_girsanov_reweighted_transition_counting.py
Comment (clonker) [inline tests/markov/test_girsanov_reweighted_transition_counting.py:1]: perhaps you can parametrize the tests to cover dense and sparse representations

**Review by clonker — CHANGES_REQUESTED** [review]
The formatting is a bit off for some of the changes (like no spaces around equal signs) and you have a few problems in your doctests.

**Comment by anyschaefer** [comment]
> If you update your branch against the current main, the errors unrelated to this PR should be gone. :)

Unfortunately, I still have the same problem. Not only the tests in  'tests/'  also the doctests gives me only message pointing to the import error.
My way to setup is: create anaconda environment with python 3.11; clone my updated fork or the original deeptime repository; in the deeptime folder with 'pip install .' install everything and with 'pytest tests/' or 'pytest --doctest-modules deeptime' I still get the import errors for the cpython extension files (e.g. 'deeptime.numeric._numeric_bindings') even if I can find them in the respective directories.
Do you know where the error can come from?

**Comment by clonker** [comment]
this is probably due to the way python discovers modules: first in the local directory. as there is a `deeptime` directory in your working dir, it'll find that one but without the compiled sources which are in your site packages :)

one way around this is using nox: `pip install nox` and then you just call `nox` inside the deeptime directory. that should take care of things. alternatively, just `cd tests` and then run
```bash
python -m pytest --doctest-modules --durations=20 --pyargs ..
```

**Comment by clonker** [comment]
ah damn, i didn't implement numpy 2.0 support yet, hence the failures. i'll push a fix soon, then ci should be green (plusminus the test you're adding).

**Review by clonker — APPROVED** [review]

**Comment by clonker** [comment]
for all i am concerned this can be squashed and merged now. i still think the testing could be a bit more thorough (as indicated by coverage) but i'll leave that up to you if you want to provide it.

File: deeptime/markov/tools/estimation/api.py [hunk 7]
```diff
@@ -213,7 +214,85 @@ def effective_count_matrix(dtrajs, lag, average='row', mact=1.0, n_jobs=None, ca
...
+    Use the reweighting approach as
+
+    >>> C_sliding = girsanov_reweighted_count_matrix(dtraj[:-1], tau, reweighting_factors)
+    
+    The generated matrix is a sparse matrix in CSR-format. For
+    convenient printing we convert it to a dense ndarray.
+
+    >>> C_sliding.toarray()
+    array([[1., 2.],
+           [0., 1.]])
```
Comment (clonker) [inline deeptime/markov/tools/estimation/api.py:284]: Sorry to bug about this again, but the test expectation does not match the result here, please do fix :)

**Review by clonker — CHANGES_REQUESTED** [review]

File: deeptime/markov/tools/estimation/api.py (hunk 7 above)
Comment (clonker) [reply deeptime/markov/tools/estimation/api.py:284]: Are you able to reproduce the test failure locally?

Comment (anyschaefer) [reply deeptime/markov/tools/estimation/api.py:284]: oh no, of course I'll fix it! unfortunately not until September.

**Comment by clonker** [comment]
Welcome back! It looks good to me now and the test failure is unrelated, so if you think it is ready, I will hit the merge button.

**Pull request merged**